### PR TITLE
snap: workaround for snapctl crash in plugin hook

### PR DESCRIPTION
**Pull request by alexzorin**
For #8922.

----

This implements a workaround similar to what is suggested at https://bugs.launchpad.net/snapd/+bug/1933392/comments/5 but this is a little bit simpler. 

The idea is that it will prioritize using `snapctl` from the `core` snap if that's available, otherwise from the`snapd` snap if it exists, and finally falling back to whatever it can find in the `PATH` (the old behavior).

I am mindful that this might go sideways, if on some systems,`snapctl` on the host works but the one in `core`/`snapd` doesn't, but it doesn't seem like a realistic concern.

I tested this on Debian Stretch and CentOS 7 (with and without installing the core snap first).

### CODE REVIEW

**Comment by ohemorange** [comment]
> I am mindful that this might go sideways, if on some systems, `snapctl` on the host works but the one in `core`/`snapd` doesn't, but it doesn't seem like a realistic concern.

We could decrease the probability of this concern by testing the execution of the various snapctls instead of just their existence; I don't know if that's necessary though.

Would you mind additionally testing on Debian 10 and current RHEL/CentOS... Stream? I think if those pass as well, we don't need to worry about doing any more stringent test in the code.

My main worry would be that this change makes us use snapctl from a place that sees more frequent updates, but seeing as how that seems to be what fundamentally caused this change anyway, I'm not actually worried about that.

**Comment by alexzorin** [comment]
Good point, testing success of `$(/path/to/snapctl get x)` would be a more reliable. 

I had the option reserved but had a sort of mindset about wanting to avoid talking to snapd as much as technically possible, to avoid talking to snapd/be faster/whatever. They're probably pretty irrational worries here so let's go with this change.

**Comment by alexzorin** [comment]
Okey dokey, seems to work OK on Debian 9, 10, CentOS 7, 8 and Stream.

File: snap/hooks/prepare-plug-plugin
Comment (ohemorange) [inline snap/hooks/prepare-plug-plugin:10]: How did you test this? When I run `snap get certbot x` on the command line, it exits 1 because we haven't set key `x`. Is it documented somewhere that `snapctl` has different behavior?

Comment (alexzorin) [reply snap/hooks/prepare-plug-plugin:10]: > How did you test this? 

One way:

    $ sudo snap run --shell certbot
    root@x1:~/devel/certbot# snapctl get x

    root@x1:~/devel/certbot# echo $?
    0

But mainly I tested it by trying out the resulting snap.

>Is it documented somewhere that snapctl has different behavior?

No, I didn't find any documentation on this one way or the other.

https://snapcraft.io/docs/using-snapctl does mention that it uses exit codes for the `is-connected` subcommand, but nothing for `get`/`set`.

Do you have another idea in mind for probing the functionality of `snapctl`? If not, I could post on the snapcraft forum and ask about this.

Comment (ohemorange) [reply snap/hooks/prepare-plug-plugin:10]: No, I like it, I was just worried by the disparity with `snap get`. Very weird that they'd differ; thanks for the explanation, lgtm.

Comment (ohemorange) [reply snap/hooks/prepare-plug-plugin:10]: Actually, now that I've said that, I would prefer using documented behavior, which doesn't seem to exist. If we post on the forums to ask, perhaps that answer would be canonical and I'd feel comfortable relying on it. But that being said, I think I'd rather merge this to get the fix in for tomorrow's release, and ask later and change it if we need to.

Comment (alexzorin) [reply snap/hooks/prepare-plug-plugin:10]: Asked at https://forum.snapcraft.io/t/is-the-exit-code-0-of-snapctl-get-some-nonexistent-option-stable/25841.

**Review by ohemorange — APPROVED** [review]

File: snap/hooks/prepare-plug-plugin
Comment (alexzorin) [reply snap/hooks/prepare-plug-plugin:10]: There has been a response: https://forum.snapcraft.io/t/is-the-exit-code-0-of-snapctl-get-some-nonexistent-option-stable/25841/3.

**Comment by bmw** [comment]
I just read through this PR and the linked threads to get caught up on this issue. The problem is the snapd maintainers are saying that the exit codes for `snapctl get` should not be relied upon at https://forum.snapcraft.io/t/is-the-exit-code-0-of-snapctl-get-some-nonexistent-option-stable/25841.

My initial thoughts on a solution here are either:

1. Go back for just checking the existence of the various snapctl executables as was done in the first version of this PR.
2. Briefly describe this issue to Ian Johnson who is the snapd maintainer commenting in both the launchpad and snapcraft thread and ask for their advice on how to handle the problem.

**Comment by ohemorange** [comment]
Personally I think (2) is the more robust long-term option, given what alex pointed out above about working vs. non-working versions, but if you haven't seen the [last update to the bug](https://bugs.launchpad.net/snapd/+bug/1933392/comments/8), I think we're not super likely to get that and given that, the safest thing is to go with option (1). Which we should probably do on the sooner side, in case they do decide to suddenly change that exit code.

**Comment by alexzorin** [comment]
> Which we should probably do on the sooner side, in case they do decide to suddenly change that exit code.

I think I'm not *super* worried about timing, because by the way it is written today, the worst thing that can happen is that Debian 9 users will find themselves unable to install DNS plugins again.  But I will try get something out for review by tomorrow or (your) Monday.

**Pull request merged**
